File: Python/Algorithm/Quick_Sort.py
# Complexity: O(n*log(n))

def partition(ar, lo, hi):
    pivot, i, j = ar[lo], lo, hi
    while i < j:
        while i < len(ar) and ar[i] <= pivot: i += 1
        while ar[j] > pivot: j -= 1
        if i < j: ar[i], ar[j] = ar[j], ar[i]
    ar[lo], ar[j] = ar[j], ar[lo]
    return j
def qsrt(ar, lo = 0, hi = None):
    if hi is None: hi = len(ar) - 1
    if lo < hi:
        pi = partition(ar, lo, hi)
        qsrt(ar, lo, pi - 1); qsrt(ar, pi + 1, hi)

File: CPP/Algorithm/Quick_Sort.cpp
// Complexity: O(n * log(n)); Worst Case: O(n^2)

int n, aa[N];
int partition(int l, int r) {
	int pivot = aa[l], i = l, j = r;
    while (i < j) {
        while (aa[i] <= pivot) ++i;
        while (aa[j] > pivot) --j;
        if (i < j) swap(aa[i], aa[j]);
    }
    swap(aa[l], aa[j]); return j;
}
void qsrt(int l = 0, int r = n - 1) {
	if (l < r) {
		int pi = partition(l, r);
		qsrt(l, pi - 1); qsrt(pi + 1, r);
	}
}